### Predict if cryptocurrencies are affected by 24-hour or 7-day price changes



In [2]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [3]:
# Load the data into a Pandas DataFrame
dfMarketData = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
dfMarketData.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [4]:
# Generate summary statistics
dfMarketData.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [5]:
# Plot your data to see what's in your DataFrame
dfMarketData.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [6]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
scaledMarketData = StandardScaler().fit_transform(dfMarketData)

In [7]:
# Create a DataFrame with the scaled data
dfScaledMarketData = pd.DataFrame(scaledMarketData)

# Copy the crypto names from the original DataFrame
dfScaledMarketData.columns = dfMarketData.columns.values

# Set the coin_id column as index
dfScaledMarketData.index = dfMarketData.index

dfScaledMarketData.dropna(inplace=True)

# Display the scaled DataFrame
dfScaledMarketData.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Scaled DataFrame.

In [8]:
# Create a list with the number of k-values from 1 to 11
ks = range(1,12)

# Create an empty list to store the inertia values
inertias = []

# Create a for loop to compute the inertia with each possible value of k
for k in ks: 
    model = KMeans(n_clusters=k, random_state=23)
    model.fit(dfScaledMarketData)
    inertias.append(model.inertia_)

In [9]:
# Create a DataFrame with the data to plot the Elbow curve
dfElbow = pd.DataFrame({'k':ks, 'inertia':inertias})
dfElbow.head()

k     inertia
0  1  287.000000
1  2  212.123342
2  3  123.190482
3  4   79.022435
4  5   68.685015

In [10]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
elbowPlot1 = dfElbow.hvplot.line(
    x='k',
    y='inertia',
    title='which value of k is best for our k-means model?',
    xticks=k
)
elbowPlot1

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** 4

---

### Cluster Cryptocurrencies with K-means Using the Original Scaled DataFrame

In [51]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4, random_state=23)

In [52]:
# Fit the K-Means model using the scaled DataFrame
model.fit(dfScaledMarketData)

KMeans(n_clusters=4, random_state=23)

In [53]:
# Predict the clusters to group the cryptocurrencies using the scaled DataFrame
clusters = model.predict(dfScaledMarketData)

# Print the resulting array of cluster values.
clusters

array([0, 0, 3, 3, 0, 0, 0, 0, 0, 3, 3, 3, 3, 0, 3, 0, 3, 3, 0, 3, 3, 0,
       3, 3, 3, 3, 3, 3, 0, 3, 3, 3, 2, 0, 3, 3, 1, 3, 3, 3, 3],
      dtype=int32)

In [54]:
# Create a copy of the scaled DataFrame
dfClusters = dfScaledMarketData.copy()

In [55]:
# Add a new column to the copy of the scaled DataFrame with the predicted clusters
dfClusters['clusters'] = clusters

# Display the copy of the scaled DataFrame
dfClusters.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  clusters  
coin_id                                             
bitcoin                        -0.251637         0  
ethereum                       -0.199352         0  
tether                         -0.282061         3  
ripple                         -0.295546         3  
bitcoin-cash                   -0.270317         0

In [56]:
# Create a scatter plot using hvPlot
clusterPlot1 = dfClusters.hvplot.scatter(
    # set `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`
    x='price_change_percentage_24h',
    y='price_change_percentage_7d',
    # color the graph points with the labels found using K-Means and
     by='clusters',
    # add the crypto name in the `hover_cols` parameter to identify
    # the cryptocurrency represented by each data point.
    hover_cols='coin_id'
)
clusterPlot1

:NdOverlay   [clusters]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [57]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3, random_state=23)

In [58]:
# Use the PCA model with `fit_transform` to reduce the original scaled DataFrame
# down to three principal components.
pcaMarket = pca.fit_transform(dfScaledMarketData)

# View the scaled PCA data
pcaMarket[:10]

array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527],
       [-0.51653377,  1.38837748,  0.80407131],
       [-0.45071134,  0.51769912,  2.84614316],
       [-0.34559977,  0.72943939,  1.47801284],
       [-0.64946792,  0.43216514,  0.60030286],
       [-0.75901394, -0.20119979, -0.21765292]])

In [59]:
# Retrieve the explained variance to determine how much information
# can be attributed to each principal component.
pca.explained_variance_ratio_

array([0.3719856 , 0.34700813, 0.17603793])

In [60]:
# Get total variance
round(sum(pca.explained_variance_ratio_), 3)

0.895

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** ~89.5% of the total variance is caused by the 3 principal components

In [61]:
# Create a new DataFrame with the PCA data.
dfPCA = pd.DataFrame(pcaMarket, columns=['pca1','pca2','pca3'])

# Copy the crypto names from the original scaled DataFrame & set as index
dfPCA.index = dfScaledMarketData.index

# Display the scaled PCA DataFrame
dfPCA.head()

pca1      pca2      pca3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

---

### Find the Best Value for k Using the Scaled PCA DataFrame

In [62]:
# Create a list with the number of k-values from 1 to 11
ks = range(1,12)

# Create an empty list to store the inertia values
inertias = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list
for k in ks:
    model = KMeans(n_clusters=k, random_state=23)
    model.fit(dfPCA)
    inertias.append(model.inertia_)

In [63]:
# Create a DataFrame with the data to plot the Elbow curve
dfElbowPCA = pd.DataFrame({'k':ks, 'inertia':inertias})
dfElbowPCA.head()

k     inertia
0  1  256.874086
1  2  182.339530
2  3  134.317570
3  4   65.275352
4  5   55.325182

In [64]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
elbowPlot2 = dfElbowPCA.hvplot.line(
    x='k',
    y='inertia',
    xticks=k
)
elbowPlot2

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** Probably 4, could also use 6


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** The original elbow also gave us 4, but there was no additional bend at 6

### Cluster Cryptocurrencies with K-means Using the Scaled PCA DataFrame

In [65]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4, random_state=23)

# Fit the K-Means model using the PCA data
model.fit(dfPCA)

# Predict the clusters to group the cryptocurrencies using the scaled PCA DataFrame
pcaClusters = model.predict(dfPCA)

# Print the resulting array of cluster values.
pcaClusters

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 2, 0, 0, 0, 2, 0,
       2, 0, 0, 0, 0, 2, 0, 2, 0, 0, 3, 0, 2, 0, 1, 2, 0, 0, 2],
      dtype=int32)

In [66]:
# Create a copy of the scaled PCA DataFrame
dfPcaClusters = dfPCA.copy()

# Add a new column to the copy of the PCA DataFrame with the predicted clusters
dfPcaClusters['cluster'] = pcaClusters

# Display the copy of the scaled PCA DataFrame
dfPcaClusters.head()

pca1      pca2      pca3  cluster
coin_id                                            
bitcoin      -0.600667  0.842760  0.461595        0
ethereum     -0.458261  0.458466  0.952877        0
tether       -0.433070 -0.168126 -0.641752        0
ripple       -0.471835 -0.222660 -0.479053        0
bitcoin-cash -1.157800  2.041209  1.859715        0

In [67]:
# Create a scatter plot using hvPlot by setting
# `x="PC1"` and `y="PC2"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
clusterPlot2 = dfPcaClusters.hvplot.scatter(
    x='pca1',
    y='pca2',
    by='cluster',
    hover_cols='coin_id'
)
clusterPlot2

:NdOverlay   [cluster]
   :Scatter   [pca1]   (pca2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [68]:
# Composite plot to contrast the Elbow curves
elbowPlot1 * elbowPlot2

:Overlay
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [69]:
# Composite plot to contrast the clusters
clusterPlot1 + clusterPlot2

:Layout
   .NdOverlay.I  :NdOverlay   [clusters]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [cluster]
      :Scatter   [pca1]   (pca2,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:**  Both of my cluster plots use clusters.I did try testing it with 3 & 6 clusters, but it did not seem to fit this data very well. Interestingly, regardless of how many clusters I used, ethlend & celsius-degree-token are in their own clusters every time.